kroxylicious-schema-tools/kroxylicious-maven-plugin/src/main/java/io/kroxylicious/CompileSchemaMojo.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.maven.plugin.AbstractMojo;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+import org.apache.maven.project.MavenProject;
+
+import io.kroxylicious.tools.schema.compiler.FatalException;
+import io.kroxylicious.tools.schema.compiler.SchemaCompiler;
+
+@Mojo(name = "compile-schema", defaultPhase = LifecyclePhase.GENERATE_SOURCES)
+public class CompileSchemaMojo extends AbstractMojo {
+
+    @Parameter(defaultValue = "${project}")
+    private MavenProject project;
+
+    /**
+     * The input file or directory to be used for generating sources
+     *
+     */
+    @Parameter(property = "kroxylicious.schema.source", defaultValue = "${basedir}/src/main/schema")
+    File source;
+
+    @Parameter(property = "kroxylicious.schema.header")
+    File headerFile;
+
+    @Parameter(property = "kroxylicious.schema.dest", defaultValue = "${basedir}/target/generated-sources/java")
+    File target;
+
+    @Parameter(defaultValue = "false")
+    boolean failOnWarnings = false;
+
+    @Parameter(property = "kroxylicious.schema.existingClasses", required = false)
+    Map<String, String> existingClasses = null;
+
+    @Override
+    public void execute() throws MojoExecutionException {
+        getLog().info("" + this);
+        String header = readHeaderFile();
+        try {
+
+            SchemaCompiler schemaCompiler = new SchemaCompiler(List.of(source.toPath()), null, header, existingClasses != null ? existingClasses : Map.of());
+            var inputs = schemaCompiler.parse();
+            var units = schemaCompiler.gen(inputs);
+            schemaCompiler.write(target.toPath(), units);
+            if (schemaCompiler.numFatals() > 0) {
+                throw new MojoExecutionException("Schema compilation failed with " + schemaCompiler.numFatals() + " fatal errors.");
+            }
+            else if (schemaCompiler.numErrors() > 0) {
+                throw new MojoExecutionException("Schema compilation failed with " + schemaCompiler.numErrors() + " errors.");
+            }
```

Comment:
Whats the practical difference between a fatal and an error?

Reply:
* Fatal means the tool isn't able to continue (it's more like an internal error). So it prevents the reporting of other problems. An example would be one of the visitors throwing an exception, or an IO exception when reading a referenced schema.
* Error is for something which is definitely wrong, perhaps because the schema is malformed (e.g. two schemas with the same `id`). We can carry on and generate something, but we know that the resulting code won't work.
* Warning is for something which is not wrong, but the user ought to know about. An example is where we ignore a file because it has a missing or unexpected `$schema` property.